WordPress/gutenberg
# Block Bindings: Improve how the context needed by sources is extended in the editor

## What?
Stealing the code from Ella in [this other pull request](https://github.com/WordPress/gutenberg/pull/60826).

This pull request uses the code from https://github.com/WordPress/gutenberg/pull/63470  to reuse and merge the `usesContext` defined by the source in the server.

There are a few things happening:
* By default, it reuses the `usesContext` defined by the source in the server registration.
* It adds a new `usesContext` property to the client source registration to let sources extend that if needed.
* It merges both and removes duplicates.
* It changes the way bindings access the context to ensure it is only available when an attribute is connected to that specific source.

## Why?
Block bindings shouldn't have to affect the block registration and the context can be accessed just in time.

Apart from that, it should reuse the server definition to avoid duplication in server and client registrations.

## Testing Instructions
I added some automated tests to ensure everything works as expected: [link](https://github.com/WordPress/gutenberg/compare/try/register-sources-from-the-server...try/improve-how-bindings-context-is-extended-in-the-editor?expand=1#diff-d5aba57655b6c1abcbd8f937a61ad98110057a6ae09f2db596811dca31c54030).
